macaw-dump/src/Data/Macaw/Dump/Discover.hs

```diff
@@ -0,0 +1,160 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Data.Macaw.Dump.Discover
+  ( DiscoverException
+  , DiscoverConfig(..)
+  , discoverConfig
+  , discover
+  ) where
+
+import Control.Exception qualified as X
+import Control.Lens qualified as Lens
+import Control.Monad qualified as Monad
+import Data.ByteString qualified as BS
+import Data.ElfEdit qualified as EE
+import Data.Foldable qualified as F
+import Data.Set qualified as Set
+import Data.Macaw.Architecture.Info qualified as MAI
+import Data.Macaw.CFG qualified as MC
+import Data.Macaw.Dump.CLIUtils qualified as MDCU
+import Data.Macaw.Discovery qualified as MD
+import Data.Macaw.Memory.ElfLoader qualified as MEL
+import Data.Macaw.Memory.LoadCommon qualified as MML
+import Data.Macaw.Memory qualified as MM
+import Data.Macaw.Symbolic qualified as MS
+import Data.Map qualified as Map
+import Data.Parameterized.Some (Some(Some))
+import Data.Text.Encoding.Error qualified as Text
+import Data.Text.Encoding qualified as Text
+import Data.Text qualified as Text
+import Lang.Crucible.Analysis.Postdom qualified as CAP
+import Lang.Crucible.CFG.Core qualified as CCC
+import Lang.Crucible.CFG.Extension qualified as CCE
+import Lang.Crucible.FunctionHandle qualified as CFH
+import Options.Applicative qualified as Opt
+import Prettyprinter qualified as PP
+import System.IO qualified as IO
+import What4.FunctionName qualified as WF
+import What4.ProgramLoc qualified as WPL
+
+data DiscoverException = ELFResolutionError String
+  deriving Show
+
+instance X.Exception DiscoverException
+
+
+-- | Convert machine addresses into What4 positions.
+--
+-- When possible, we map to the structured 'WPL.BinaryPos' type. However, some
+-- 'MM.MemSegmentOff' cannot be mapped to an absolute position (e.g., some
+-- addresses from shared libraries are in non-trivial segments). In those cases,
+-- we map to the unstructured 'WPL.Others' with a sufficiently descriptive string.
+--
+-- TODO: import from `Testing`
```

Comment:
Does this TODO still need to be resolved?

Reply:
Eh, it's a pretty small function, I'm not too worried about this level of duplication on second thought :shrug: